ldurniat/berry.lua

```diff
@@ -447,484 +468,553 @@ end
-				object = objects[j]
-
-				-- Make sure we have a properties table
-				object.properties = object.properties or {}
+				map:createObject(object, layer)
 
-				-- Image/Sprite	
-				-- GID stands for global tile ID
-				if object.gid then
-
-					-- Original code from https://github.com/ponywolf/ponytiled    
-				    flip.x  = hasbit( object.gid, FlippedHorizontallyFlag )
-				    flip.y  = hasbit( object.gid, FlippedVerticallyFlag )          
-				    flip.xy = hasbit( object.gid, FlippedDiagonallyFlag )
+			end
 
-				    object.gid = clearbit( object.gid, FlippedHorizontallyFlag )
-				    object.gid = clearbit( object.gid, FlippedVerticallyFlag )
-				    object.gid = clearbit( object.gid, FlippedDiagonallyFlag )
+		elseif layer.type == 'tilelayer' then
 
-					-- Get the correct tileset using the GID
-					tileset = getTilesetFromGID( object.gid, tilesets )
+			-- layer.width variable is already taken due to layer being a display group, 
+			-- so we need a new variable name (layer.size works in this case)
+			layer.size = info.width
 
-					if tileset then
+			for position, gid in ipairs(info.data) do -- GID stands for global tile ID
 
-						local firstgid,   tileId      = tileset.firstgid,  object.gid - tileset.firstgid
-						local width,      height      = object.width, object.height
-						local imageSheet, pathToImage = loadTileset( tileset, tilesetsDirectory, tileId ) 
+				if gid > 0 then map:createTile(position, gid, layer) end
 
-						if imageSheet then
+			end
 
-							if findProperty( layer.properties, 'isAnimated' ) or findProperty( object.properties, 'isAnimated' ) then
+		end
 
-								image = display.newSprite( objectLayer, imageSheet, tileset.sequenceData )
+		map:insert( layer )
 
-							else 
+	end 
 
-								image = display.newImageRect( objectLayer, imageSheet, tileId + 1, width, height )
+	-- Add useful properties
+    map.defaultExtensions = 'berry.plugins.'
+    map.tilewidth         = data.tilewidth
+    map.tileheight        = data.tileheight
+    map.designedWidth     = data.width  * data.tilewidth
+    map.designedHeight    = data.height * data.tileheight
 
-							end
-								
-						else image = display.newImageRect( objectLayer, pathToImage, width, height ) end
+    -- Center map
+    map.x, map.y = display.contentCenterX - map.designedWidth * 0.5, display.contentCenterY - map.designedHeight * 0.5
 
-						local points, x, y, rotation  = retrieveShapeData( tileId , tileset )
+	-- Set the background color to the map background
+	display.setDefault( 'background', decodeTiledColor( data.backgroundcolor ) )   
+	return map
+end
 
-						-- Add collsion shape
-						if points then
+------------------------------------------------------------------------------------------------
+--- Create and add tile to layer
+--  
+------------------------------------------------------------------------------------------------
+function Map:createTile(position, gid, layer)
```

Comment:
Do you test/use it? I'm afraid it can be too complicted to use. The same apply for `:createObject`.

Add `:clone()` function for duplicate objects/tiles can be very useful:) What you think?

Reply:
I ripped out the code for the tile and object creation and put them into their own functions for a few reasons.  My primary concern was the ability to reuse these functions to create tiles and objects both before and after the map was loaded.  (this makes content creation more dynamic)

Another reason was mainly code readability.  Before my changes, the map creation function was nested about 6 loops in depth.  This was buried too deep and is generally not good coding standard or practice.  If a function is getting to be too big it's usually a good idea to break it up into smaller functions, which is what I did.  On top of that, careful consideration needs to be made with the scopes of the variables invovled.  

So before my changes the function for the map creation looked like this:

```lua
function Map:new( filename, tilesetsDirectory )
	-- blah blah
	for layer do
		-- blah blah
		if layer.type == 'objectgroup' then
				if object.gid then
					-- blah
					if tileset then
						-- blah
						if imageSheet then
							if findProperty( layer.properties, 'isAnimated' ) or findProperty( object.properties, 'isAnimated' ) then
								-- blah
							end
						end	

						if points then
							if #points > 8 * 2 then
								-- blah
							end
						end
					end	

				elseif object.polygon or object.polyline then 
					-- blah
					for i = 1, #points do
						-- blah
			    	end

				    if object.polygon then 
						if #points > 8 * 2 then 
							-- blah
						end		
				    end
				end

				if image then
					-- blah
					if flip.xy then
						-- blah
					end

					if  findProperty( object.properties, 'hasBody' ) then 
						-- blah
					end	
					-- blah
				end	
			end

		elseif layer.type == 'tilelayer' then
			-- blah
			if gid > 0 then
				-- blah
				if tileset then
					-- blah
					if imageSheet then
						-- blah
					end	

					if image then
						-- blah
					end	
					-- blah
				end	
			end		
		end	
	end 
end
```   

and now it's been simplified to this:

```lua
function Map:new( filename, tilesetsDirectory )
	-- blah blah

	for layer do
		-- blah blah
		if layer.type == 'objectgroup' then
			-- blah blah
			for _, object in ipairs(objects) do
				map:createObject(object, layer)
			end

		elseif layer.type == 'tilelayer' then
			-- blah blah
			for position, gid in ipairs(info.data) do
				if gid > 0 then 
					map:createTile(position, gid, layer) 
				end
			end
		end
	end 
end
```   

Much easier to read and understand IMO.  And it's only nested 3-4 loops deep.  

I believe the `createTile` function might be better suited towards being set as local function instead of being public.  The only reason I made it so, was because I was following the same process as `createObject` when I was refactoring the code.  I doubt many users will want to generate tiles as a tile layer after the map has been loaded.  It is too complex and diffcult to utilize, especially with the postiion and gid.  

For `createObject` on the other hand, this is a different scenario.  Especially if there is TexturePacker support!  The only things a user should need to add the object to a map would be the x and y position and the sprite name.  This would make it simple to add objects to the game.  No dealing with gid's or row/column positions.  Just load the texturepacker sprites into a imagesheet table somewhere and boom.  Ready to go!  At least that's how I'm envisioning it being implemented.

Adding a `:clone()` function might be useful and shouldn't be too difficult to add.  Will probably only be good for cloning objects though as using it on a tile or tilelayer could get complex.

Reply:
Ok. Thanks for explanation.